tools/setup-envtest/main.go

```diff
@@ -0,0 +1,292 @@
+
+	// printFormat is the flag value for -p, --print
+	printFormat = envp.PrintOverview
+	// zapLvl is the flag value for logging verbosity
+	zapLvl = zap.WarnLevel
+
+	binDir       = flag.String("bin-dir", "", "directory to store binary assets (default: $OS_SPECIFIC_CACHE_DIR/envtest-binaries)")
+	remoteBucket = flag.String("remote-bucket", "kubebuilder-tools", "remote GCS bucket to download from")
+	remoteServer = flag.String("remote-server", "storage.googleapis.com", "remote server to query from.  You can override this if you want to run an internal storage server instead, or for testing.")
+)
+
+// TODO: handle interrupts?
+
+const (
+	defaultBinDir = "envtest-binaries"
+)
+
+// getUseEnv gets the envUseEnv env var, or returns "never"
+// as a default if it's not set
+func getUseEnv() string {
+	fromEnv := os.Getenv(envUseEnv)
+	if fromEnv != "" {
+		return fromEnv
+	}
+	return "never"
+}
+
+// setupLogging configures a Zap logger
+func setupLogging() logr.Logger {
+	logCfg := zap.NewDevelopmentConfig()
+	logCfg.Level = zap.NewAtomicLevelAt(zapLvl)
+	zapLog, err := logCfg.Build()
+	if err != nil {
+		envp.ExitCause(1, err, "who logs the logger errors?")
+	}
+	return zapr.NewLogger(zapLog)
+}
+
+// setupEnv initializes the environment from flags.
+func setupEnv(globalLog logr.Logger, version string) *envp.Env {
+	log := globalLog.WithName("setup")
+	if *binDir == "" {
+		cacheDir, err := os.UserCacheDir()
+		if err != nil {
+			envp.ExitCause(1, err, "unable to deterimine default binaries directory (use --bin-dir to manually override)")
+		}
+
+		*binDir = filepath.Join(cacheDir, defaultBinDir)
+	}
+	log.V(1).Info("using binaries directory", "dir", *binDir)
+
+	env := &envp.Env{
+		BinDir: *binDir,
+		Log:    globalLog,
+		Client: &remote.Client{
+			Log:    globalLog.WithName("storage-client"),
+			Bucket: *remoteBucket,
+			Server: *remoteServer,
+		},
+		VerifySum:     *verify,
+		ForceDownload: *force,
+		NoDownload: *installedOnly,
+		Platform: versions.PlatformItem{
+			Platform: versions.Platform{
+				OS:   *targetOS,
+				Arch: *targetArch,
+			},
+		},
+		FS: afero.Afero{Fs: afero.NewOsFs()},
+		Store: store.NewAt(*binDir),
+		Out: os.Stdout,
+	}
+
+	switch version {
+	case "", "latest":
+		env.Version = versions.LatestVersion
+	case "latest-on-disk":
+		env.Version = versions.AnyVersion
+		env.NoDownload = true
+	default:
+		var err error
+		env.Version, err = versions.FromExpr(version)
+		if err != nil {
+			envp.ExitCause(1, err, "version be a valid version, or simply 'latest' or 'latest-on-disk'")
+		}
+	}
+
+	env.CheckCoherence()
+
+	return env
+}
+
+func main() {
+	// exit with appropriate error codes -- this should be the first defer so
+	// that it's the last one executed.
+	defer envp.HandleExitWithCode()
+
+	// set up flags
+	flag.Usage = func() {
+		name := os.Args[0]
+		fmt.Fprintf(os.Stderr, "Usage: %s [FLAGS] use|list|cleanup|sideload [VERSION]\n", name)
+		flag.PrintDefaults()
+		fmt.Fprintf(os.Stderr,
+			`
```

Comment:
WDYT about making this tool `alpha` for now so we can get feedback on usage before defining a stable CLI?
```suggestion
			`
**Note:** this command is currently alpha, and it's CLI may change.

```